modules/core/src/lib/attribute.js

```diff
@@ -176,47 +187,57 @@ export default class Attribute extends BaseAttribute {
         this.buffer.reallocate(this.value.byteLength);
       }
 
-      state.needsUpdate = true;
+      if (state.updateRanges !== range.FULL) {
+        this.value.set(oldValue);
+        // TODO - copy on the GPU
+        this.buffer.subData(oldValue);
+      }
+
+      this.setNeedsUpdate(true, {startRow: instanceCount});
       state.allocedInstances = allocCount;
       return true;
     }
 
     return false;
   }
 
-  updateBuffer({numInstances, bufferLayout, data, startRow, endRow, props, context}) {
+  updateBuffer({numInstances, bufferLayout, data, props, context}) {
     if (!this.needsUpdate()) {
       return false;
     }
 
     const state = this.userData;
 
-    const {update} = state;
+    const {update, updateRanges} = state;
 
     let updated = true;
     if (update) {
       // Custom updater - typically for non-instanced layers
-      update.call(context, this, {data, startRow, endRow, props, numInstances, bufferLayout});
+      for (const [startRow, endRow] of updateRanges) {
+        update.call(context, this, {data, startRow, endRow, props, numInstances, bufferLayout});
+      }
       if (this.constant || !this.buffer || this.buffer.byteLength < this.value.byteLength) {
         // Full update
         this.update({
```

Comment:
This branch calls a helper method but the other branch does the update inline. Maybe align (e.g. use helper methods in both branches, or have the full update just prepare a range that covers all and fall through)?

Reply:
`update` is a method of the `BaseAttribute` class that uploads the attribute value to the GPU. It's certainly confusing given the various things called `update` here.